### DISCUSSION TITLE
Cursor-based pagination without sequential cursor.

### DISCUSSION THREAD

**Discussion by lskuper** [prisma/prisma#4888] [Q&A] [answered]
Hi everyone, 

for Cursor-based pagination the docs https://www.prisma.io/docs/concepts/components/prisma-client/pagination says:
"A cursor bookmarks your location in a result set and **must be a unique, sequential column** - such as an ID or a timestamp."

 My tests work although the cursor is **not sequential**.  
 I achieve cursor pagination behavior by using a unique but nonsequential key as a cursor (e.g. a UUID) and outgoing from this cursor orderBy a sequential key (which not has to be unique).  
 
 The query looks like:
```ts
 const res = await prismaClient.post.findMany({
  cursor: { id: currentCursor.id },
  take: 5,
  orderBy: { createdAt: "asc" },
});  
// res contains the objects as expected: 
// It contains 5 objects sequential by createdAt starting the object the cursor pointed to.
```

Schema:
```prisma
 model Post { 
   id        String   @id @default(uuid()) 
   createdAt DateTime @default(now()) 
 }
```

I assume the docs are not correct here, and the cursor does not have to be sequential, right?

**Comment by mhwelander** [comment, marked as answer]
Hello and thanks for an interesting question!

<s>To clarify, **unique** in this context does not have to mean a field explicitly marked `@unique` in the schema- just a field that is extremely likely to have unique values, like a blog post timestamp. I will make sure we make this clear in the docs.</s>

Edit: I've been told that ordering by a field that is not explicitly unique (literally has a `@unique` attribute) is possible, **but** results in a less efficient query and performance degradation after the first page ([more info here](https://github.com/prisma/prisma/issues/4490#issuecomment-752985665)). If you aren't Amazon and can more or less guarantee that `createdAt` will be unique, make it explicitly `@unique` in your schema. 

Your particular example works because your timestamps field values are unique (technically) and sequential, and your cursor is unique. You're paginating on a kind of 'super-cursor' made up of a timestamp and an ID. You would get exactly the same results if you used `createdAt` as a cursor rather than `id`.

In addition, there are two things that could go wrong in your version (although both are pretty unlikely to be honest):

## Scenario 1: Someone changes a blog post date

The following issue is only likely to happen if:

- Your dataset is very dynamic - stuff is being added, edited, backdated, and removed all the time
- Your dataset is being paginated through very often by a lot of people

Here's what could go wrong:

1. You are paginating through the data set. Your next cursor is **DA4** with the timestamp **10-01-21:**

![image](https://user-images.githubusercontent.com/849648/103755747-98e76d80-500e-11eb-903d-d0dc2e47d09d.png)

     You're expecting the next page too includes A12, S99 and so on.

2. But.. *while* you're still looking at the old page, someone **changes the date on DA4 (your cursor):**

![image](https://user-images.githubusercontent.com/849648/103755769-a3a20280-500e-11eb-95ac-144d8ae0cebd.png)

3. You try to go to the next page using DA4 as a cursor. But, because the date has changed on DA4 and you are ordering by date, you don't get A12, S99, and so on - you get data you've already seen:

![image](https://user-images.githubusercontent.com/849648/103755837-c3d1c180-500e-11eb-8d2b-c42558109881.png)

This would not happen if you used a timestamp as a cursor. Your cursor would be **10-01-21** instead of DA4, and no matter what changes are made in the dataset, you will get data *after* 10-01-21.

## Scenario 2:  False sense of security if timestamps aren't unique

If your timestamps truly are **not unique** (for example, you store only dates), then using ID as a cursor does not help.

In the following example, every single post has the same date and we're using ID as a cursor:

![image](https://user-images.githubusercontent.com/849648/103755878-d64bfb00-500e-11eb-8de1-905d8eb35aa5.png)

It might *seem* like pagination is working, but the sort order on identical values (06-01-21 in this case) **is unpredictable**. Sometimes it could be the order of rows in the table, sometimes the order could come from an index. I might use DA4 as a cursor and find that its position in the results has changed, and get results I've already seen.

To conclude - it is safest to use a unique, sequential field and order by that same field. Hope that helps, let me know if you have follow-up questions!

**Reply by lskuper** [reply]
Thanks for the detailed reply, it definitely helped me! 
I wasn't aware of the both Scenarios you pointed out!

> To clarify, unique in this context does not have to mean a field explicitly marked @unique in the schema- just a field that is extremely likely to have unique values, like a blog post timestamp.

I cannot confirm this, prismaClient forces me to assign an object to cursor that is `PostWhereUniqueInput` (which totally makes sense). 

To sum up my understanding:
1) the cursor must be a @unique column in the model.
2) the cursor is not necessarily the key to order by. orderBy can be _theoretically_ (considering the possible ) any arbitrary key.
3) the orderBy key should be unique and sequential (but does not need to be marked as @unique in the schema). The order of objects that have the same value in the orderBy key, is unpredictable (!).

For my specific case that leads to the conclusion that if I 
a) make sure that the value of the key I orderBy never changes (--> Scenario 1 cannot happen),
b) accept the risk of undefined order in Scenario 2 because I estimate it for my case as extremely unlikely,

the query
```
const res = await prismaClient.post.findMany({
  cursor: { id: currentCursor.id },
  take: 5,
  orderBy: { createdAt: "asc" },
});  
```
does the job, right?

Cheers

**Reply by mhwelander** [reply]
Hi again! No problem, this has been a real brain workout for me today 😄 

> the cursor must be a `@unique` column in the model.

Yes, you're right, my mistake - I should have said the column you `orderBy` needs to be unique (to avoid scenario 2) but does not necessarily need to be marked `@unique`'. However, I spoke too soon here - [it seems that you can run into performance issues](https://github.com/prisma/prisma/issues/4490#issuecomment-752985665) sorting by field not explicitly marked `@unique`.

> the cursor must be a @unique column in the model.

Correct, as you said.

> the cursor is not necessarily the key to order by. orderBy can be theoretically (considering the possible ) any arbitrary key.

Correct - if it's unique and sequential, as you say below. (Note for myself: I guess one reason to use a different `cursor` and `orderBy` is if you get the `cursor` (an ID) from your UI and nothing else, and then `orderBy` some other field like `position` or `sequence` or `createdBy`.) I will update the docs - this is a great point I didn't think about.

> the orderBy key should be unique and sequential (but does not need to be marked as @unique in the schema). The order of objects that have the same value in the orderBy key, is unpredictable (!).

Technically yes, but not sorting on a unique field may result in performance issues at scale - see https://github.com/prisma/prisma/issues/4490#issuecomment-752985665. _If_ you notice this issue, some solutions include:

* In your case you might be able to mark `createdBy` as `@unique` (if your project is to create blog posts) 
* Alternatively, if you only expect to do shallow pagination (not jump to page 100,000), switch to offset pagination if the performance benefit of cursor pagination is undermined by not being able to sort on a `@unique` field. 
* If you truly need to paginate through a large data set and need cursor pagination, consider creating a dedicated `position` column (unique, sequential) and using that to sort

To conclude - yes your example works fine; scenarios 1 and 2 are unlikely to happen and you might not notice performance issues until later, but at that point you have options to try.

**Reply by joshuaavalon** [reply]
Sorry for replying an old discussion but after I read through the discussion, I came up with some questions.

In lskuper's example, sorting by `createdAt` does not guarantee a stable sort but we can sort by multiple columns.

```js
const res = await prismaClient.post.findMany({
  cursor: { id: currentCursor.id },
  take: 5,
  orderBy: [{ createdAt: "asc" }, { id: "asc" }]
});
```

By fallback to sort by `id`, we can guarantee a stable sort. 

For the performance issue, I think that is because database automatic create index on unique column. If we create an index on `createdAt` and `id`, shouldn't it solve the problem?

**Reply by ezzabuzaid** [reply]
Hi @joshuaavalon, is the `id` sequential here?

**Comment by mhwelander** [comment]
Docs issue to incorporate points from discussion: https://github.com/prisma/docs/issues/1146

**Reply by sandbox-apps** [reply]
What would be the best strategy for this? By just reading the discussions my query seems like a mistake to be seen

```ts
this.prisma.groupContent.findMany({
      where: {
        groupId: param.groupId,
      },
      take: 100,
      orderBy: {
        hotIndex: 'desc',
      },
    });
```

As hotIndex (is the relevance algorithm based on the likes/votes/reaction) so in short this will be always changing.
We dont want to use a createdAt for social media as it is just sorted by latest...
We need to sort for same key metrics e.g. like count which we also assume that it will be changing and changing?

Does this mean I cannot use cursor pagination here and instead go back to `skip` instead of `cursor`?

**Reply by ryands17** [reply]
You have two ways to go about this:
1. You need to have both `hotIndex` and `id` in sort as mentioned in the above comment to guarantee a stable sort.
2. Use offset pagination

**Reply by sandbox-apps** [reply]
Thanks @ryands17 seems easy to go with option 1. As I love the idea of the cursor as it really perfectly fits mobile first applications... 

https://github.com/prisma/prisma/discussions/4888#discussioncomment-403826

**Comment by rtpa25** [comment]
Hey I was trying to make a cursor with the help of my createdAt field in the DB as the Prisma docs mentions that we can make use of both id and timestamps in cursors but I am unable to make it work with createdAt would really appreciate any sort of help

**Reply by ratasorin** [reply]
If the `createdAt` field is `@unique` in your schema
```prisma
createdAt DateTime @unique @default(now())
```
You can use the `createdAt` field as a cursor because the database can locate the last record you queried (based on the enforced uniqueness of dates) and can reliably move up and down records (based on the natural sequentiality of dates).

As an example: 
```js
const records = await prisma.posts.findMany({
  cursor: {
   createdAt: lastCursor.createdAt
  },
  orderBy: {
   createdAt: 'asc', // to go from old to new
   createdAt: 'desc' // to go from new to old 
 },
  where: {
   createdAt: {
    gt: lastCursor.createdAt, // to take newer records than the last one (gt = greater than)
    lt: lastCursor.createAt // to take older records than the last one (lt = less than)
  }
  },
  limit: 100
});
```

**Comment by mikevercoelen** [comment]
Hello everyone, consider the following schema:

```prisma
model Channel {
  id            String    @id @default(cuid())
}

model Event {
  id            String    @id @default(cuid())
  createdAt     DateTime  @default(now())
}
```

We're building cursor based pagination to get all events from a channel, for our business this is an absolutely crucial endpoint so has to be solid.

This is what we have now:

```typescript
const eventsFromChannel = await prisma.event.findMany({
  take: limit + 1,
  skip: 1,
  cursor: {
    id: cursor
  },
  where,
  orderBy: {
    createdAt: 'desc'
  }
})
```

Few problems with this:

- `createdAt` is not unique, therefore it might lead to problems (we have heavy volume)
- Both `channel.id` and `event.id` are strings / cuid's, so is this solid for cursors?

What we have considered is:

Adding a `idIncremented` to both `Channel` and `Event` models, and on `Event` create a `position Int` field that is a compound field.

Help would be much appreciated!

**Reply by ratasorin** [reply]
I think that an `eventSequence` field on the `Event` is a very good idea if `createdAt` cannot be unique and the `id` cannot be sequential.
```prisma
eventSequence Int @unique @default(autoincrement())
```
This can also free you from needing to `orderBy: id`, as the records are already marked as `@unique`.

**Reply by cesalberca** [reply]
@mikevercoelen did you manage to solve this? I'm facing the same issue. Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
